### ISSUE TITLE
Recipe datagen compatibility issue (conditional & special recipe combination)

### ISSUE THREAD

**Issue by desht** [neoforged/NeoForge#223]
**Minecraft Version:** 1.20.2

**NeoForge Version:** 20.2.3-beta

**Logs:** n/a

**Steps to Reproduce:**
 1. See description

**Description of issue:**

In Forge on 1.20.1, I could do this in my recipe datagen:
```java
ConditionalRecipe.builder()
                .addCondition(new ModLoadedCondition("patchouli"))
                .addRecipe(c -> SpecialRecipeBuilder.special(ModRecipes.GUIDE_BOOK.get()).save(c, RL("guide_book").toString()))
                .build(consumer, RL("guide_book"));
```
to generate this:
```json
{
  "type": "forge:conditional",
  "recipes": [
    {
      "conditions": [
        {
          "type": "forge:mod_loaded",
          "modid": "patchouli"
        }
      ],
      "recipe": {
        "type": "modularrouters:guide_book",
        "category": "misc"
      }
    }
  ]
}
```

However, `ConditionalRecipe` has changed quite a bit in 1.20.2:
* No static `builder()` method, but simply constructing it appears to work - is this correct?
* More seriously, `ConditionalRecipe#withRecipe()` now takes a `RecipeBuilder` argument. While this is OK for shaped/shapeless recipes, it doesn't work for special recipes, i.e. the `SpecialRecipe` class, since `SpecialRecipeBuilder` does not implement `RecipeBuilder` but instead extends the abstract `CraftingRecipeBuilder` class.

So I'm not sure if I'm just doing it wrong, or if Neo needs a bit of extra support here for wrapping special recipes inside conditional recipes?

**Comment by Technici4n** [comment]
It seems that a few other porting mistakes where made when porting this class:
- It is not possible to have multiple recipes in a conditional recipe anymore. Previously one could generate multiple recipes, each with its own condition. This is indeed the whole point of having the conditional recipe type in the first place.
- The advancement generation options are gone. Previously there was a way to generate a corresponding advancement with `generateAdvancement()`, or to provide a custom conditional advancement with `setAdvancement`.
- Speaking of which, conditional advancements are now gone. However it is important to disable the advancement associated with a conditional recipe.

It will be annoying and quite inconvenient to add conditions to the `Advancement` record, but I'm not sure we have the choice. We can easily patch in an `Optional` field only used for data generation.

Another option is that we associate conditions with objects via a global identity hash map. See:
- https://github.com/FabricMC/fabric/blob/3f301502811f0ab410f045d065b6f98d5b630537/fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/impl/datagen/FabricDataGenHelper.java#L218
- https://github.com/FabricMC/fabric/blob/3f301502811f0ab410f045d065b6f98d5b630537/fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricRecipeProvider.java#L67
- https://github.com/FabricMC/fabric/blob/3f301502811f0ab410f045d065b6f98d5b630537/fabric-data-generation-api-v1/src/main/java/net/fabricmc/fabric/api/datagen/v1/provider/FabricRecipeProvider.java#L97

**Comment by Technici4n** [comment]
Fixed in PR #250. While you can use the new API, note that we will mostly likely be changing the conditions key from `conditions` to `neoforge:conditions` shortly, so you will have to re-run datagen at some point once that lands.
